lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -0,0 +1,538 @@
+            if isinstance(e, OpenSearchPluginRelationClusterNotReadyError):
+                self.charm.status.set(WaitingStatus("s3-changed: cluster not ready yet"))
+            else:
+                self.charm.status.set(
+                    BlockedStatus("Unexpected error during plugin configuration, check the logs")
+                )
+                # There was an unexpected error, log it and block the unit
+                logger.error(e)
+            event.defer()
+            return
+
+        if self._plugin_status not in [
+            PluginState.ENABLED,
+            PluginState.WAITING_FOR_UPGRADE,
+        ]:
+            self.charm.status.set(WaitingStatus(f"Plugin in state {self._plugin_status}"))
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            # Plugin is configured locally for this unit. Now the leader proceed.
+            self.charm.status.clear(ActiveStatus())
+            return
+
+        self.apply_api_config_if_needed()
+        self.charm.status.clear(ActiveStatus())
+
+    def apply_api_config_if_needed(self) -> None:
+        """Runs the post restart routine and API calls needed to setup/disable backup.
+
+        This method should be called by the charm in its restart callback resolution.
+        """
+        # Backup relation has been recently made available with all the parameters needed.
+        # Steps:
+        #     (1) set up as maintenance;
+        self.charm.status.set(MaintenanceStatus("Configuring backup service..."))
+        #     (2) run the request; and
+        state = self._register_snapshot_repo()
+        #     (3) based on the response, set the message status
+        if state == BackupServiceState.SUCCESS:
+            self.charm.status.set(ActiveStatus("Backup service configured"))
+        else:
+            self.charm.status.set(BlockedStatus(f"Backup setup failed with {state}"))
+        self.charm.status.clear(ActiveStatus())
+
+    def _on_s3_broken(self, event: EventBase) -> None:
+        """Processes the broken s3 relation.
+
+        It runs the reverse process of on_s3_change:
+        1) Check if the cluster is currently taking a snapshot, if yes, set status as blocked
+           and defer this event.
+        2) If leader, run API calls to signal disable is needed
+        """
+        if self.charm.model.get_relation(S3_RELATION).units:
+            # There are still members in the relation, defer until it is finished
+            event.defer()
+            return
+        self.charm.status.set(MaintenanceStatus("Disabling backup service..."))
+        snapshot_status = self._check_snapshot_status()
+        if snapshot_status in [
+            BackupServiceState.SNAPSHOT_IN_PROGRESS,
+        ]:
+            # 1) snapshot is either in progress or partially taken: block and defer this event
+            self.charm.status.set(
+                BlockedStatus(f"Disabling backup not possible: {snapshot_status}")
```

Comment:
Maybe?
```suggestion
                MaintanceStatus(f"Disabling backup postponed till backup in progress: {snapshot_status}")
```
can be separate PR